File: backend/app/routers/authors.py
from fastapi import APIRouter, Depends, HTTPException, status
from typing import Annotated
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.db.session import get_db
from app.models.author import Author
from app.models.entry import EntryRevision, EntryTargetType
from app.models.user import User
from app.schemas.author import AuthorCreate, AuthorRead, AuthorUpdate
from app.schemas.entry import EntryRevisionRead
from app.security import require_admin_user
from app.services.entries import create_author_entry, update_author_entry

router = APIRouter(prefix="/authors", tags=["authors"])


@router.get("", response_model=list[AuthorRead])
def list_authors(db: Annotated[Session, Depends(get_db)], limit: int = 50, offset: int = 0):
    statement = select(Author).order_by(Author.sort_name, Author.name).limit(limit).offset(offset)
    return db.scalars(statement).all()


@router.post("", response_model=AuthorRead, status_code=status.HTTP_201_CREATED)
def create_author(
    payload: AuthorCreate,
    current_user: Annotated[User, Depends(require_admin_user)],
    db: Annotated[Session, Depends(get_db)],
    change_note: str = "",
):
    return create_author_entry(db, payload, changed_by=current_user, change_note=change_note)


@router.get("/{author_id}", response_model=AuthorRead)
def get_author(author_id: int, db: Annotated[Session, Depends(get_db)]):
    author = db.get(Author, author_id)
    if author is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Author not found")
    return author


@router.patch("/{author_id}", response_model=AuthorRead)
def update_author(
    author_id: int,
    payload: AuthorUpdate,
    current_user: Annotated[User, Depends(require_admin_user)],
    db: Annotated[Session, Depends(get_db)],
    change_note: str = "",
):
    return update_author_entry(db, author_id, payload, changed_by=current_user, change_note=change_note)


@router.get("/{author_id}/history", response_model=list[EntryRevisionRead])
def list_author_history(author_id: int, db: Annotated[Session, Depends(get_db)]):
    if db.get(Author, author_id) is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Author not found")

    statement = (
        select(EntryRevision)
        .where(
            EntryRevision.entity_type == EntryTargetType.AUTHOR,
            EntryRevision.entity_id == author_id,
        )
        .order_by(EntryRevision.created_at.desc(), EntryRevision.id.desc())
    )
    return db.scalars(statement).all()

File: frontend/src/api/authors.ts
import { apiRequest } from './client'
import type { EntryRevision } from './suggestions'

export interface Author {
  id: number
  name: string
  sort_name?: string | null
  bio: string
  birth_year?: number | null
  death_year?: number | null
  created_at: string
  updated_at: string
}

export interface CreateAuthorInput {
  name: string
  sort_name: string | null
  bio: string
  birth_year: number | null
  death_year: number | null
}

export async function listAuthors(limit = 500, offset = 0): Promise<Author[]> {
  const params = new URLSearchParams({
    limit: String(limit),
    offset: String(offset),
  })

  return apiRequest<Author[]>(`/api/v1/authors?${params.toString()}`, {
    fallbackErrorMessage: 'Unable to load authors.',
  })
}

export async function getAuthor(authorId: string | number): Promise<Author> {
  return apiRequest<Author>(`/api/v1/authors/${authorId}`, {
    fallbackErrorMessage: 'Unable to load this author.',
  })
}

export type UpdateAuthorInput = Partial<CreateAuthorInput>

export async function createAuthor(
  payload: CreateAuthorInput,
  accessToken?: string,
  changeNote = '',
): Promise<Author> {
  const params = new URLSearchParams()
  if (changeNote.trim()) {
    params.set('change_note', changeNote.trim())
  }

  return apiRequest<Author>(`/api/v1/authors${params.size ? `?${params.toString()}` : ''}`, {
    accessToken,
    method: 'POST',
    body: payload,
    fallbackErrorMessage: 'Unable to add this author.',
  })
}

export async function updateAuthor(
  authorId: string | number,
  payload: UpdateAuthorInput,
  accessToken?: string,
  changeNote = '',
): Promise<Author> {
  const params = new URLSearchParams()
  if (changeNote.trim()) {
    params.set('change_note', changeNote.trim())
  }

  return apiRequest<Author>(`/api/v1/authors/${authorId}${params.size ? `?${params.toString()}` : ''}`, {
    accessToken,
    method: 'PATCH',
    body: payload,
    fallbackErrorMessage: 'Unable to update this author.',
  })
}

export async function listAuthorHistory(authorId: string | number): Promise<EntryRevision[]> {
  return apiRequest<EntryRevision[]>(`/api/v1/authors/${authorId}/history`, {
    fallbackErrorMessage: 'Unable to load author history.',
  })
}
